### PR TITLE
[WIP] Upgrade to TS 3.0

### PR DESCRIPTION
**Pull request by blink1073**
Takes advantage of the new [project references](https://blogs.msdn.microsoft.com/typescript/2018/07/30/announcing-typescript-3-0/#project-references) feature to simplify how we build our packages.

Everything Just Works ™️ in that each folder gets its own `lib/` as if it were built directly, huzzah! However, because the output of metapackage itself has changed, we need to update the way we build our API docs. We need to build docs in each folder and make our own index page. No code has to change, and Typedoc uses its own TypeScript compiler version, so we're good to go. This will actually be an improvement since our current API landing page is awkward.

- [x] Use project references
- [x] Update our integrity script to populate `references`
- [x] Clean up `watch` mode in light of these simplifications
- [x] Fix handling of `update:dependency foo@next`
- [x] Audit all `watch` scripts - do not use `-b` in examples since they should be standalone
- [x] Clean up API docs handling
- [x] Add top level landing page

Updated status: still blocked on a TypeScript Issue: https://github.com/Microsoft/TypeScript/issues/26955

### CODE REVIEW

**Comment by SimonBiggs** [comment]
Ahh this is awesome. Thanks @blink1073!

**Comment by jasongrout** [comment]
Awesome, thanks!

I'm thinking we should maybe abandon Typedoc. It seems the project has stalled, and it is preventing us from using newer Typescript syntax (there are a few places noted in the code that we'd like to use new syntax).

One question is: how much are the api docs being used? Perhaps we should add a google analytics tracker to give us an idea. (plus it might be interesting to see which api pages are most used...)

**Comment by jasongrout** [comment]
This was way easier than I thought. I thought we would be adding a references section to each package, keeping it in sync with the dependencies for the packages. We might eventually do that, but for now this makes the transition nice and smooth.

**Comment by SimonBiggs** [comment]
@jasongrout if by typedoc you mean the following:

https://jupyterlab.github.io/jupyterlab/globals.html

I can let you know that personally I initially found it quite confusing, and in the end I found following definitions within visual studio code and reading source code to be easier.

**Comment by jasongrout** [comment]
> if by typedoc you mean the following:
> 
> https://jupyterlab.github.io/jupyterlab/globals.html

Yes, that's what I mean. Thanks for your feedback!

**Comment by blink1073** [comment]
@simonbiggs, was it confusing because of the way it was structured?  If we had a better way to get at a class definition page, would that help?

**Comment by SimonBiggs** [comment]
I guess traversal through the code base by using control + click within
visual studio code allowed me to quickly work out what was going on.
Everything was commented so well and in a consistent manner (I suspect for
use within typedoc) that all the information to use the API was available
there. If I ever needed more help I could jump out of the type definition
and go to the source code without much extra effort.


Being able to navigate that way, directly as I am writing code, is hard to
beat.


When I was new to the code base, typedoc was the only online documentation at
the time. Given that it was my first port of call to learn JupyterLab it
was overwhelming. The experience for new commers now might be quite a bit
different given that they would use the primary online docs first, and once
they are ready, might then use the API docs.


Personally I found, to understand what was really going on initially, I
really had to read through examples of how the API was being used within
the source code. I would find something I wanted to use, then run a search
for that method within the source code and be provided with examples of you
guys using it. I would sometimes just pull out my phone and scout through
the source code on GitHub on my phone. What I could learn by doing that was
far beyond what I felt I could learn looking at the typedoc.

**Comment by blink1073** [comment]
Okay, this *will* work, but we do need to do what you thought @jasongrout, maintain the appropriate `references` for each project.  That will be a part of the `integrity` check.  The bonus thing is, you can then run `yarn build` from *any* package at any time, and it will Just Work :tm:.

**Comment by jasongrout** [comment]
yay for incremental compilation! I suppose `--watch` will work nicely as well?

**Comment by blink1073** [comment]
It sounds that way, I won't know for sure until build works.

**Comment by blink1073** [comment]
Drat, we need to wait on this [problem](https://github.com/Microsoft/TypeScript/issues/26147) in watch mode.  Otherwise, it works great!

**Comment by blink1073** [comment]
It takes ~1s to build on my machine when there are no changes, and ~73s from scratch.

**Comment by blink1073** [comment]
But the amazing thing is that you can run `yarn build` *from any package* and it will Just Work :tm:

**Comment by saulshanabrook** [comment]
Yay, this is wonderful! thanks so much for working on this.

I am in favor of removing typedoc. I have never used it while getting acquainted with the codebase and like @SimonBiggs found it much easier to hop around it using vscode.

**Comment by saulshanabrook** [comment]
Am I understanding things right that tsconfig.json references should match package.json references to other jupyterlab packages? I know that typescript has an issue for making it easier to have the references be inferred from other files, but in the meantime is there any checking to make sure those stay in sync?

File: examples/console/src/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  "extends": "../../../tsconfigbase",
+  "compilerOptions": {
+    "outDir": "../build"
+  },
+  "include": ["*"],
```
Comment (saulshanabrook) [inline examples/console/src/tsconfig.json]: Can these `"include": ["*"],` options be moved to `tsconfigbase`?

**Comment by blink1073** [comment]
The references are populated in our ensure-package script that is run in the integrity script.

File: examples/console/src/tsconfig.json (hunk 1 above)
Comment (blink1073) [reply examples/console/src/tsconfig.json]: Sounds good

Comment (blink1073) [reply examples/console/src/tsconfig.json]: Nope, I tried and the files don't get built.

**Comment by blink1073** [comment]
Here are all the times we reference a class definition in our docs:

```
$ search "http://jupyterlab.github.io/jupyterlab/classes"
./developer/extension_dev.rst:91:`shell <http://jupyterlab.github.io/jupyterlab/classes/_application_src_shell_.applicationshell.html>`__
./developer/documents.rst:25:`Document widgets <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html>`__ represent a view of a document model. There can be multiple document widgets associated with a single document model, and they naturally stay in sync with each other since they are views on the same underlying data model.
./developer/documents.rst:29:Registry <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html>`__
./developer/documents.rst:35:Manager <http://jupyterlab.github.io/jupyterlab/classes/_docmanager_src_manager_.documentmanager.html>`__
./developer/documents.rst:51:`Widget Factories <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html#addwidgetfactory>`__
./developer/documents.rst:60:`Model Factories <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html#addmodelfactory>`__
./developer/documents.rst:68:`Widget Extension Factories <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html#addwidgetextension>`__
./developer/documents.rst:80:`File Types <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html#addfiletype>`__
./developer/documents.rst:86:`File Creators <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html>`__
./developer/notebook.rst:22:`NotebookWidgetFactory <http://jupyterlab.github.io/jupyterlab/classes/_notebook_src_widgetfactory_.notebookwidgetfactory.html>`__
./developer/notebook.rst:24:`NotebookPanel <http://jupyterlab.github.io/jupyterlab/classes/_notebook_src_panel_.notebookpanel.html>`__
./developer/notebook.rst:37:`NotebookModel <http://jupyterlab.github.io/jupyterlab/classes/_notebook_src_model_.notebookmodel.html>`__
./developer/notebook.rst:82:   widget <http://jupyterlab.github.io/jupyterlab/classes/_notebook_src_widget_.notebook.html>`__.
./developer/notebook.rst:99:`NotebookActions <http://jupyterlab.github.io/jupyterlab/classes/_notebook_src_actions_.notebookactions.html>`__
./developer/notebook.rst:112:   `InputArea <http://jupyterlab.github.io/jupyterlab/classes/_cells_src_inputarea_.inputarea.html>`__,
./developer/notebook.rst:115:      `CodeEditorWrapper <http://jupyterlab.github.io/jupyterlab/classes/_codeeditor_src_widget_.codeeditorwrapper.html>`__,
./developer/notebook.rst:120:`CodeCell <http://jupyterlab.github.io/jupyterlab/classes/_cells_src_widget_.codecell.html>`__
./developer/notebook.rst:122:`OutputArea <http://jupyterlab.github.io/jupyterlab/classes/_outputarea_src_widget_.outputarea.html>`__.
./developer/notebook.rst:124:`OutputAreaModel <http://jupyterlab.github.io/jupyterlab/classes/_outputarea_src_model_.outputareamodel.html>`__
./developer/notebook.rst:126:`RenderMimeRegistry <http://jupyterlab.github.io/jupyterlab/classes/_rendermime_src_registry_.rendermimeregistry.html>`__
./developer/notebook.rst:266:Registry <http://jupyterlab.github.io/jupyterlab/classes/_docregistry_src_registry_.documentregistry.html>`__.
```

My vote is to keep TypeDoc for now.   There is an open (passing) PR to upgrade to 3.0: https://github.com/TypeStrong/typedoc/pull/825/files.

**Comment by SimonBiggs** [comment]
Having typedoc as a supporting document for the official docs themselves
makes sense.

**Comment by jasongrout** [comment]
> My vote is to keep TypeDoc for now. There is an open (passing) PR to upgrade to 3.0: https://github.com/TypeStrong/typedoc/pull/825/files.

That was fast. There have been open PRs updating TS to other TS versions in the past months that went nowhere, though, so I'm not exactly holding my breath.

**Comment by blink1073** [comment]
I'm not one to sit on my hands :wink:.  https://github.com/TypeStrong/typedoc/issues/826#issuecomment-410995038

**Comment by ellisonbg** [comment]
Are we still going to try to get this in for 0.34?

**Comment by blink1073** [comment]
We're blocked on fix in TypeScript.  I'll leave this marked as 0.34 in case the fix is available in time.

**Comment by blink1073** [comment]
Update: the fix to TypeScript is done in master and will be backported to 3.0.2.

**Comment by jasongrout** [comment]
For reference, the TS issue was https://github.com/Microsoft/TypeScript/issues/26147, and the PRs fixing it are linked in that issue.

**Comment by blink1073** [comment]
Update: we are no longer blocked!  I'll return to this after #5251.

```
[04:37:29] Starting compilation in watch mode...

[04:37:29] Found 0 errors. Watching for file changes.
```

**Comment by blink1073** [comment]
@jasongrout, why was `--listEmittedFiles` needed?  I see you added it in https://github.com/jupyterlab/jupyterlab/pull/4703, but I can't puzzle out the reason it was needed.  I ask because `tsc --build` does not support that flag.  Also, we are blocked again by Microsoft/TypeScript#26955.

**Comment by blink1073** [comment]
Okay, this is all done pending Microsoft/TypeScript#26955.

**Comment by blink1073** [comment]
Rock: meet hard place: https://github.com/TypeStrong/ts-loader/issues/815.   I'll remove `--build` support in our tests for now, until we finish the Jest conversion...

File: .lintstagedrc [hunk 2]
```diff
@@ -1,7 +1,8 @@
 {
   "linters": {
     "**/*{.ts,.tsx,.js,.jsx,.css,.json,.md}": [
-      "prettier --write"
+      "prettier --write",
+      "git add"
```
Comment (vidartf) [inline .lintstagedrc]: Will this prevent partial staging/commit of a file?

Comment (blink1073) [reply .lintstagedrc]: The file was by definition already staged, this makes sure that the changes are included in the commit.

Comment (vidartf) [reply .lintstagedrc]: You can do partial staging with e.g. `git add -p` or `git gui`, where you stage chunks/lines. I think running `git add` automatically will override any such partial staging.

Comment (vidartf) [reply .lintstagedrc]: C.f. https://github.com/okonet/lint-staged/issues/62

Comment (blink1073) [reply .lintstagedrc]: I'm not stepping into that landmine.  This change only corrects on omission that brings it in line with the other two linters in this file.

Comment (vidartf) [reply .lintstagedrc]: Sure, I just wonder if there are some way for me to disable these actions? I use partial staging a lot, so I'd rather learn which commands to run!

Comment (blink1073) [reply .lintstagedrc]: I don't know how to handle that case

Comment (vidartf) [reply .lintstagedrc]: I'll try to open up another PR for it.

Comment (blink1073) [reply .lintstagedrc]: Great!

**Comment by jasongrout** [comment]
> @jasongrout, why was `--listEmittedFiles` needed?

I'm not sure if it was *needed*, though I found it a nice confirmation - if I switch to the watch window, I can confirm that the file I was working on was explicitly built. If it's messing up things, feel free to take it back off.

**Comment by blink1073** [comment]
There's an open issue to add support for the rest of the flags for `-w` when using `-b`, but it isn't there yet.    https://github.com/Microsoft/TypeScript/issues/25613

**Pull request merged**